### PR TITLE
ECT: owner-unique cross-box borrowing and seam-synchronized conformal B push

### PR DESCRIPTION
**Pull request by clarkse-he**
## Summary

Rescoped per the review discussion below: this PR now contains only the cross-box
correctness core (3/3 of the split). The convergence tests are #7153, the atomic area
accounting is #7152 (this PR is based on it — the first commit in the diff is #7152's; marked
draft until #7152 merges), and the opt-in balanced borrow will follow separately.

Each fab previously ran the face-extension passes over its own copies of faces on
shared nodal planes, so a physical face's borrowing decision could be made twice (once
per adjacent fab) from possibly-stale ghost data, and the lent-area accounting never
propagated back to the owner. The enlarged-cell gather/scatter in the conformal B push
similarly read and wrote neighbor faces across box seams without reconciliation.

## What's changed

- **Owner-unique borrowing decisions.** Borrowing decisions are made only by the owner
  copy of each face (`amrex::OwnerMask`, rebuilt on init, restart and regrid). The area
  each lender gave away and the intruded-face marks are accumulated per direction and
  reduced to the owners between the extension passes (`SumBoundary` of the remote
  part); all copies are then made owner-consistent (`OverrideSync`) before the BCK
  correction reads them.
- **Seam-synchronized conformal B push.** With cut cells at fab seams the borrowed Venl
  contributions scatter across boxes: `EvolveBCartesianECT` now defers the
  unstable-face B update until the scattered contributions have been summed to the
  owners (`SumBoundary` + `OverrideSync` of Venl between an assembly phase and an
  update phase). Single-box non-periodic layouts keep the fused, communication-free
  path bit-identically. `EvolveRhoCartesianECT` refreshes one ghost layer of the face
  circulations at seams so the enlarged-cell gather reads owner-computed values.
- **Regrid:** the ECT flag fields (`flag_info_face`, `flag_ext_face`) are remade on
  regrid alongside the borrowing structure.
- **New test** `test_2d_embedded_boundary_rotated_cube_multibox`: the rotated-cube
  cavity with `amr.max_grid_size = 16` on 2 ranks, so the cut-cell region crosses box
  seams and CI exercises the cross-box path.

## Evidence

Default-path preservation (2D rotated-square cavity, 68 steps, CPU): results are
**bit-identical to `development`** in every configuration tested — single box, 16 boxes
on 1 rank, 16 boxes on 2 ranks, 128² with 16-cell boxes, and with explicit 4³ OMP
tiling (max relative difference 0.0 in all six field components).

GPU determinism (see measurement comment below): on the 4-box rotated-square cavity,
run-to-run max relative differences drop from 4.5e-1 (`development`) to 2.5e-14 (this
branch + #7152).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FyFeNNpr5jir3ygSakbZCm

### CODE REVIEW

**Comment by clarkse-he** [comment]
GPU determinism follow-up, as promised in the description.

Setup: the 2D rotated-square ECT cavity at 128², 4 boxes on a single GPU (a
box seam crossing the cut-cell region), 68 steps; three identical runs per
binary; the raw staggered fields of the final step compared across runs.

| binary | worst run-to-run max rel. difference |
|---|---|
| `development` | **4.5e-1** |
| this branch | **2.5e-14** |

The pre-change extension passes race on concurrent borrows on GPU: the
borrowing outcomes differ from run to run, and those differences amplify
through the cavity dynamics to order unity. With the atomic test-and-subtract
and owner-unique decisions, the same configuration reproduces to
accumulation-order round-off. (On CPU both are deterministic, per the
description — the races have no lanes to run in.)

**Comment by clarkse-he** [comment]
Merged `development` to resolve conflicts. One note on scope: #7100 landed the same `Gpu::Atomic::AddNoRet` fix for the Venl scatter in `EvolveBCartesianECT`, so that piece of the "atomic area accounting" bullet is now upstream and has dropped out of this PR's diff. The atomics in the face-extension passes (`WarpXFaceExtensions.cpp`) and the owner-unique borrowing remain unique to this PR.

**Comment by RemiLehe** [comment]
Thanks for this PR @clarkse-he 

It seems that this PR introduces 4 several separate functionalities (listed in summary), and as a result is rather large. Would you be able to break this into separate PRs, so as to make them easier to review?

**Comment by clarkse-he** [comment]
Done — the PR is split into four, one functionality each:

1. **Convergence tests** (new PR #7153): the 2D Yee-vs-ECT convergence chain for the
   rotated-square cavity. Pure test addition, independent of any source change (the
   asserted orders hold on stock `development`).
2. **Atomic area accounting** (new PR #7152): the atomic test-and-subtract in the two
   face-extension passes. One file, GPU-only effect; bit-identical on CPU.
3. **This PR (rescoped + retitled): owner-unique cross-box borrowing and the
   seam-synchronized conformal B push** — the multi-box/GPU determinism core, now based
   on #7152 (marked as draft until #7152 merges; the first commit in the diff is #7152's). The
   GPU run-to-run determinism evidence in the comment above (4.5e-1 → 2.5e-14) was
   measured with exactly the content of #7152 + this PR. A new
   `test_2d_embedded_boundary_rotated_cube_multibox` test (2 ranks, box seams crossing
   the cut-cell region) makes CI exercise the cross-box path.
4. **Balanced borrow** (`warpx.eb_ect_balanced_borrow`): deferred; I will open it with
   its own checksum test after #7152 and this PR land, to keep the review queue short.

The `neigh_faces → neighbor_faces` rename is dropped from all four to keep the diffs
minimal.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 1]
```diff
@@ -525,7 +634,40 @@ WarpX::ComputeFaceExtensions ()
...
+        for (int idim = 0; idim < 3; ++idim) {
+            auto& marks = *intruded_mark[idim];
+            marks.SumBoundary(0, 1, marks.nGrowVect(), amrex::IntVect(0), period);
+            auto* info_mf = m_flag_info_face[maxLevel()][idim].get();
+            for (amrex::MFIter mfi(marks); mfi.isValid(); ++mfi) {
+                const amrex::Box bx = mfi.validbox();
+                auto const& mk = marks.const_array(mfi);
+                auto const& info = info_mf->array(mfi);
+                amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                    if (mk(i, j, k) > 0 && info(i, j, k) == 1) { info(i, j, k) = 2; }
```
Comment (aarontran) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]:
```suggestion
                    if (mk(i, j, k) > 0 && info(i, j, k) == FaceInfo::available) {
                        info(i, j, k) = FaceInfo::intruded;
                    }
```
using enum in recent dev commit 2cf00a17

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 2]
```diff
@@ -665,6 +819,7 @@ WarpX::ComputeOneWayExtensions ()
                     const amrex::Real S_stab = ::ComputeSStab(i, j, k, lx, ly, lz, dx, dy, dz, idim);
 
                     const amrex::Real S_ext = S_stab - S(i, j, k);
+                    int n_borrowed = 0;
```
Comment (aarontran) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]:
```suggestion
```
Removed by previous PR #7152 commit 0986772 .  But it may not get cleaned up by a `git merge BLAST-WarpX:development` invocation (at least it didn't for my local repo at cmd line)

**Comment by aarontran** [comment]
Few suggestions.  It may help to merge or cherry-pick `2daeef5116` and `0986772dc7` from PR #7152 into this feature branch.  When using the sole command `git merge BLAST-WarpX:development`, I saw that a few changes were dropped by the automatic git-merge algorithm.

Once that's done: to make the `ComputeEightWaysExtensions` all-or-nothing rollback work, seems at least two changes are needed (flagged by Claude Opus 5).

First, in the `ComputeEightWaysExtension` rollback code, add:
```
amrex::Gpu::Atomic::AddNoRet(
    ::GetNeighPtr(lent, i, j, k, vec(0), vec(1), idim),
    - borrowing_area[ps + n]);
```
Second, the `intruded_mark` data structure can be removed entirely, since its entries can be deduced from `lent_area` and its rollback code would be a bit uglier anyways.  The associated synchronization calls for `m_flag_ext_face` can also be put into `sync_lent_area` for simplicity

I'm working to test this PR a bit more, hope to have an update (or will let you know if dropped) in the next few days

**Comment by clarkse-he** [comment]
Merged rather than rebased, so the commits you reviewed are still in history. Head is now `ad9fe2b43`. Heads-up before anything else: the branch moved under you, so whatever you had in flight against `643cf59f2` is stale — sorry for the timing.

## On the merge: you were right, and the reason is worse than a lost rename

A plain `git merge development` here leaves 9 conflict hunks — 8 in `WarpXFaceExtensions.cpp`, 1 in `EvolveB.cpp`. The reason is that this branch's first commit, `934808ca3`, *was* the earlier draft of what became #7152. So the two sides of most of those hunks are that draft versus the merged all-or-nothing form, and resolving them toward HEAD doesn't just drop a rename, it reinstates superseded code. Concretely, taking the HEAD side gives you back:

* the `n_borrowed` counter in `ComputeOneWayExtensions`, instead of keying the reset off `flag_ext_face`
* the `count < nborrow` cap that can silently skip a still-available neighbour, instead of `count == nborrow` falling through to the rollback
* the entire rollback block in `ComputeEightWaysExtensions` — the all-or-nothing mechanism itself — replaced by the partial-extension bookkeeping it was written to remove
* literal flag values instead of the `FaceInfo` enumerators
* the `issue #2257 / PR #2298` comments Remi asked to drop

and taking the development side alone drops this branch's `lent_area` recording at both borrow sites.

So I resolved it against a rebase of the branch onto development — dropping `934808ca3` outright, which is the honest thing to do with a superseded draft — and then checked that the merge tree came out byte-identical to that rebase before committing it. `934808ca3` is still in the branch's history but is now inert: nothing of it survives in the tree, and the squash will take care of it.

## Your two changes

**1. The `lent_area` unwind.** Implemented, and you were right that it was a real bug rather than a tidiness issue: within a fab the restored `S_mod` covers it, but across a seam a lender owned by another fab only ever learns of a borrow through that ledger, so it was charged for area it had already been given back — breaking the `S_mod + (sum of lent areas) == S` identity that the intruded-face branch of `EvolveBCartesianECT` relies on.

I took a slightly different route than `AddNoRet(..., -borrowing_area[ps + n])`: the eight-ways pass now records its lent areas once the all-or-nothing outcome is known, rather than per patch as it borrows. Same net effect, and it avoids a `+x` then `-x` on the same address, whose residual is not exactly zero once other faces' contributions interleave between the two. `sync_lent_areas` tests `rem != amrex::Real(0.)` exactly, so an eps-sized residual would both charge and mark a lender that ended up lending nothing. Recording once keeps a rolled-back face's ledger at exactly zero. The one-way pass already records inside the successful branch, so it needed no change.

**2. `intruded_mark` is gone entirely**, as you suggested. A lender is intruded exactly when the remote part of its lent-area total is non-zero — which `sync_lent_areas` already computes in order to charge `S_mod` — so the mark now falls out of that same loop: one fused pass instead of a second `iMultiFab`, a second `SumBoundary` and a second sweep over the box, and the flag can no longer disagree with the area ledger. The `flag_info_face` and `m_flag_ext_face` synchronization moved in with it, also as you suggested.

One consequence worth flagging: marking now happens between the passes rather than after both. That changes no availability decision, because `FaceInfo::available` and `FaceInfo::intruded` are both lendable — the same argument the old deferred-marking comment was making, just running in the other direction.

Your inline suggestion is in as well. That one is worth noting as its own hazard: the reduction is new code on this branch, so no conflict was raised there and the literals would have survived the merge silently.

## Testing

2D+3D, EB, OMP, MPI. All three rotated-cube ECT tests pass run, analysis and checksum, including `test_2d_embedded_boundary_rotated_cube_multibox` on 2 ranks — whose benchmarks were committed against the explicit-mark structure, so the deduced marks reproduce it bit for bit. I have not re-run on GPU since the merge; the concurrency claims in the commit message still rest on the earlier measurements and on yours, not on a fresh run. Say the word if a GPU re-run of the multibox case would be useful to you and I'll do it.

There is also a small scrub commit (`ad9fe2b43`, no functional change): `m_ect_needs_seam_sync` was documented as though it were a per-face proximity test when it is set from the layout alone (`boxArray().size() > 1 || isAnyPeriodic()`), and the single-level assumption in `EvolveBCartesianECT` was a bare `AMREX_ALWAYS_ASSERT` reporting only the expression. That assertion is load-bearing rather than decorative, which was not obvious from the code: nothing in the tree forbids `max_level > 0` with ECT, and the owner masks are allocated at `maxLevel()` only.

Take the time you need on the testing — happy to hold the branch still if that helps.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 1 above)
Comment (clarkse-he) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: Applied, thanks — `ad9fe2b43`. Worth noting this one was a genuine hazard rather than a style nit: the mark reduction is new code on this branch, so the merge with development raised no conflict here and the literal `1`/`2` would have survived it silently, unlike the two borrow sites where the same rename did conflict.

The surrounding block has since moved into `sync_lent_areas`, since `intruded_mark` is gone and the mark is now deduced from the lent-area remote part — details in the PR comment above.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 3]
```diff
@@ -564,7 +566,91 @@ WarpX::ComputeFaceExtensions ()
...
+    // of shared nodal planes. The area each lender gave away (lent_area) is
+    // therefore accumulated alongside the direct writes and reduced to the
+    // owners between the passes; a non-zero remote contribution is also what
+    // marks a lender as intruded on its owner, so the flag and the area
+    // ledger cannot disagree. Single-box non-periodic layouts skip every
+    // reduction and keep the historical communication-free behavior
+    // bit-identically.
+    const bool multi_box = (boxArray(maxLevel()).size() > 1)
+        || Geom(maxLevel()).isAnyPeriodic();
+    m_ect_needs_seam_sync = multi_box;
```
Comment (aarontran) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]:
```suggestion
    // owns, but lending faces may be ghosts or non-owned copies of shared
    // nodal planes. The lent_area must be accumulated and then reduced to
    // owners after the 1- and 8-way passes.  Single-box non-periodic layouts
    // skip reductions.
    m_ect_needs_seam_sync = (Bfield[0]->boxArray().size() > 1
                             ||  Geom(maxLevel()).isAnyPeriodic());
```

Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: boxArray(maxLevel()).size() returned 1 when 16 was expected in a test case, causing synchronization to silently be skipped; Claude suggest it's due to order of AMReX initialization.  Accessing via Bfield[0] seems to work instead

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 4]
```diff
@@ -564,7 +566,91 @@ WarpX::ComputeFaceExtensions ()
...
+    //
+    // A face is marked intruded exactly when the remote part is non-zero.
+    // Both passes record a lent area only once the borrow is known to have
+    // succeeded, so a face that a rolled-back extension had briefly taken
+    // area from contributes exactly zero here, and is neither charged nor
+    // marked. Marking between the passes is safe because FaceInfo::available
+    // and FaceInfo::intruded are both lendable, so it changes no availability
+    // decision in the pass that follows.
+    auto const sync_lent_areas = [&] () {
+        if (!multi_box) { return; }
```
Comment (aarontran) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]:
```suggestion
        if (!m_ect_needs_seam_sync) { return; }
```

**Comment by aarontran** [comment]
With above bugfix for `m_ect_needs_seam_sync`, the cross-box sync within `WarpXFaceExtensions.cpp` looks good based on three tests shown below (which make use of PR #7245).  However, I have NOT yet reviewed the logic in `EvolveB.cpp` and `EvolveECTRho.cpp`.

This PR addresses two problems in the face extension passes:
1. Area borrowed from ghosts not added up correctly.
    + Fixed and verified by tests below.
2. Faces on shared nodal face planes: two boxes may disagree on which neighbors to borrow from to stabilize the same unstable face.
    + I don't have a test case that shows this disagreement yet, but it appears theoretically possible if concurrency in eight-way passes is not fully nailed down, such the eight-way pass rollbacks & borrowing decisions on a nodal plane are not deterministic across different AMReX boxes.
    + Code in this PR looks like it should fix the issue, but I have not studied or instrumented the relevant code paths to really verify... in the absence of any known symptoms (unless @clarkse-he  has any?) I propose to skip rigorous verification of this edge case.
    + (not sure yet whether this issue rears its head in the `EvolveB`, `EvolveECTRho` code)

20260903_ECT_box_sync_testing_test1
20260903_ECT_box_sync_testing_test2
20260903_ECT_box_sync_testing_test3

Text in test#3 above is not quite right, min(S_mod) is computed only on the faces that participate in lending and borrowing (i.e., for which S != S_mod)

File: Source/Parallelization/WarpXRegrid.cpp [hunk 5]
```diff
@@ -203,6 +203,8 @@ WarpX::RemakeLevel (int lev, Real /*time*/, const BoxArray& ba, const Distributi
                     RemakeMultiFab( m_eb_update_E[lev][idim] );
                     RemakeMultiFab( m_eb_update_B[lev][idim] );
                     if (WarpX::electromagnetic_solver_id == ElectromagneticSolverAlgo::ECT) {
+                        RemakeMultiFab( m_flag_info_face[lev][idim] );
+                        RemakeMultiFab( m_flag_ext_face[lev][idim] );
```
Comment (aarontran) [inline Source/Parallelization/WarpXRegrid.cpp:211]: This fix looks reasonable, but I've not traced detailed code architecture.  Claude Opus suggests this fix's purpose is to update the DistributionMapping for these iMultiFabs -- can an expert confirm purpose + necessity?

If yes+yes to latter question, does `m_ect_face_owner_mask` need to be similarly remade here?

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 6]
```diff
@@ -243,10 +245,26 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
 
     amrex::LayoutData<amrex::Real> *cost = WarpX::getCosts(lev);
 
+    auto& warpx = WarpX::GetInstance();
+    // With cut cells at fab seams (multi-box or periodic layouts), the
+    // enlarged-cell gather/scatter crosses boxes: the B update is then
+    // deferred until the scattered Venl contributions have been summed to
+    // the owners. Single-box non-periodic layouts take the fused
+    // communication-free path below, bit-identical to the historical one.
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    // electromotive forces contributed to Venl must scatter/gather across
    // boxes prior to the B-field update.
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 7]
```diff
@@ -243,10 +245,26 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+    const bool seam_sync = warpx.ECTNeedsSeamSync();
+    // The ECT solver is single-level: ComputeFaceExtensions fills the
+    // borrowing structure and the owner masks at maxLevel() only, so there is
+    // nothing for a coarser level to read here
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    // borrowing structure and the owner masks at maxLevel() only.
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 8]
```diff
@@ -243,10 +245,26 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(lev == warpx.maxLevel(),
+        "EvolveBCartesianECT: the ECT solver does not support mesh refinement");
+
     Venl[0]->setVal(0.);
     Venl[1]->setVal(0.);
     Venl[2]->setVal(0.);
 
+    // Phase 1: assemble the enlarged-face accumulator and scatter the
+    // borrowed contributions (owner-gated when syncing across seams);
+    // in the fused path this also applies the unstable-face B update.
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    // borrowed contributions (owner-gated when syncing across seams).
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 9]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
 
                 }
 
-                B(i, j, k) = B(i, j, k) - dt * rho_enl;
+                // In the seam-sync path the unstable-face B update is
+                // deferred to phase 2 (after the Venl reduction); the fused
+                // path applies it here, as it always has
+                if (!owner) {
+                    B(i, j, k) = B(i, j, k) - dt * rho_enl;
+                }
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: I think removing this branched code, and doing unstable face B update in the "phase 2" loop, can give a clearer code structure
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: If restructuring of Phase 1/2 loops below is accepted, then this line can be deleted as `B` will not be used in Phase 1 loop

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 10]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+            amrex::Gpu::synchronize();
+            wt = static_cast<amrex::Real>(amrex::second()) - wt;
+            amrex::HostDevice::Atomic::Add( &(*cost)[mfi.index()], wt);
+        }
+    }
+
+    // Phase 1.5: reduce the scattered borrowed contributions across fab
+    // seams (ghost entries and shared nodal-plane copies) to the owners,
+    // then make all copies bit-equal. Without seams this is skipped and the
+    // ghost-scatter behavior matches the historical single-box path.
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    // then make all copies bit-equal.
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 11]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+                *Venl[idim], 0, 1, amrex::IntVect(1), amrex::IntVect(0),
+                WarpX::do_single_precision_comms, period);
+            Venl[idim]->OverrideSync(period);
+        }
+    }
+
+    // Phase 2: apply the B updates. In the fused path only the stable faces
+    // remain; in the seam-sync path the unstable faces are updated here too,
+    // from the now-complete Venl (Venl/S_mod is the same expression the
+    // fused path evaluates as rho_enl).
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
    // Phase 2: apply the B updates.
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 12]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+
+        if (cost && WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers) {
+            amrex::Gpu::synchronize();
+        }
+        auto wt = static_cast<amrex::Real>(amrex::second());
+
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+            Array4<Real> const &B = Bfield[idim]->array(mfi);
+            Array4<Real> const &Rho = ECTRhofield[idim]->array(mfi);
+            Array4<Real> const &Venl_dim = Venl[idim]->array(mfi);
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
            amrex::Array4<Real> const &B = Bfield[idim]->array(mfi);
            amrex::Array4<Real> const &Rho = ECTRhofield[idim]->array(mfi);
            amrex::Array4<Real> const &Venl_dim = Venl[idim]->array(mfi);
```
following https://warpx.readthedocs.io/en/latest/developers/contributing.html#style-and-conventions

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 13]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
+            amrex::Array4<int> const &flag_info_cell_dim = flag_info_cell[idim]->array(mfi);
+            amrex::Array4<Real> const &S = face_areas[idim]->array(mfi);
+            amrex::Array4<Real> const &S_mod = area_mod[idim]->array(mfi);
+
+            const bool sync = seam_sync;
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp [hunk 14]
```diff
@@ -357,15 +384,74 @@ void FiniteDifferenceSolver::EvolveBCartesianECT (
...
 
                 if (flag_info_cell_dim(i, j, k) == FaceInfo::extended) {
+                    // Unstable face: deferred update of the seam-sync path
+                    // (the fused path already updated it in phase 1)
+                    if (sync) {
+                        B(i, j, k) = B(i, j, k) - dt * Venl_dim(i, j, k) / S_mod(i, j, k);
+                    }
                     return;
                 }
                 else if (flag_info_cell_dim(i, j, k) == FaceInfo::available) {
```
Comment (aarontran) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]:
```suggestion
                    // Unstable cell
                    B(i, j, k) = B(i, j, k) - dt * Venl_dim(i, j, k) / S_mod(i, j, k);
                } else if (flag_info_cell_dim(i, j, k) == FaceInfo::available) {
```

**Review by aarontran — COMMENTED** [review]
OK with me to merge this PR after the `m_ect_needs_seam_sync` bugfix above is added; other comments can be skipped if needed.

* Last batch of suggestions = cosmetic tweaks to `EvolveB` face update for readability.
* Attached test shows:
    + Fix to `m_ect_needs_seam_sync` brings eigenmode closer to analytic expectation
    + Tweaks to `EvolveB` do not introduce obvious code regression
    + WARNING: test plots = AI generated without line-by-line micromanagement
* Not reviewed (deprioritized):
    + `Examples/...` and `Regression/`, I suggest to remove the "multibox" input & associated edits entirely to save review time/hassle, since (i) checksum may now be invalid, (ii) default cube's rotation angle does not give too many unstable cut-cells near AMReX box seams which are needed to exercise this PR, (iii) purpose of "multibox" input is partly covered by this PR's review dialogue
    + Load-balance cost calculation code in `EvolveB` was not studied

Test setup = identical simulation (2D rotated-cube, theta=0.05 radians) run using four different versions of code w/ stacked edits.  Test results mention commits "f31..." and "b0e..." which aren't publicly visible; they are my local repo commits of suggested changes from this PR review.  Inputs here: [inputs_test_2d_embedded_boundary_rotated_cube_seam.txt](https://github.com/user-attachments/files/31996089/inputs_test_2d_embedded_boundary_rotated_cube_seam.txt)

Image

Image

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 3 above)
Comment (clarkse-he) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: Applied in c7e9ac100 — and the mechanism is confirmed: `AmrMesh::MakeNewGrids` calls `MakeNewLevelFromScratch` (→ `InitLevelData` → `InitializeEBGridData` → `ComputeFaceExtensions`) *before* `SetBoxArray` publishes `grids[lev]`, so `boxArray(maxLevel())` is empty during init-from-scratch. The restart and regrid paths set the BoxArray first, which is why only fresh starts were affected. The field's own BoxArray is what `init_borrowing` already keys on.

Thanks for catching this. It also means the multi-box "bit-identical to development" numbers in the description were taken with the sync inadvertently off; refreshed numbers in the summary comment below.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 4 above)
Comment (clarkse-he) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: Applied in c7e9ac100.

File: Source/Parallelization/WarpXRegrid.cpp (hunk 5 above)
Comment (clarkse-he) [reply Source/Parallelization/WarpXRegrid.cpp:211]: Yes, and yes-but-already-done. `InitializeEBGridData(lev)` is called a few lines below (line 220) and writes into both flag fields through MFIters over the new `Bfield` layout, so they must be on the new `DistributionMapping` first — that is what the `RemakeMultiFab` does. The owner mask needs no separate remake because `InitializeEBGridData` rebuilds it from scratch (the `amrex::OwnerMask` block in `WarpXInitData.cpp` right before `ComputeFaceExtensions`). Added a comment saying so in c7e9ac100.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 9 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Taken, in 463525b20. It is bit-identical on the single-box path (verified against `ad9fe2b43` on the 2D rotated cube: every `Level_0/Cell_D_*` file identical by `cmp`, all six field components bit-equal, and the single-box checksum benchmark reproduced to 0.00e+00): an `extended` face never lends, so its `Venl` entry receives no atomic adds from other iterations, and phase 2 divides the same two operands phase 1 divided for `rho_enl`.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 14 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Done in 463525b20, with one deliberate deviation: written as `dt * (Venl_dim(i,j,k) / S_mod(i,j,k))`. The un-parenthesized form groups as `(dt * Venl) / S_mod` and rounds differently from the historical `dt * rho_enl` (measured 4e-15 relative on the single-box test); the parenthesized form reproduces it exactly, and it also makes the seam-sync path agree with the single-box path bit-for-bit instead of to roundoff.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Done in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 13 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Done in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 6 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 7 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 8 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 10 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 11 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp (hunk 12 above)
Comment (clarkse-he) [reply Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp]: Applied in 463525b20.

**Comment by clarkse-he** [comment]
Pushed c7e9ac100 (seam-sync decision read off the field + regrid comment) and 463525b20 (single two-phase B update in `EvolveBCartesianECT`) per @aarontran's review.

Verification (CPU, 2D rotated cube):
- single box, 1 rank: bit-identical to `ad9fe2b43` (every `Level_0/Cell_D_*` file identical by `cmp`, all six field components bit-equal, and the single-box checksum benchmark reproduced to 0.00e+00);
- 16 boxes, 2 ranks: differs from `ad9fe2b43` by up to 1.6e-2 (By) / 1.8e-2 (Ex) / 3.7e-2 (Ez) relative, and the fixed multi-box run is byte-identical to the fixed single-box run — expected, since the seam sync is now actually active on a fresh start. That also means the multi-box bit-identity evidence in the description was collected with the sync off; the description's evidence paragraph needs refreshing accordingly.
- Faces to be extended (multi-box): 56 / 4 / 0 y-faces (initial / after one-way / after eight-way), identical single- and multi-box.

On the nodal-plane disagreement (problem 2): we have no observed symptom, and given the above the earlier multi-box runs say nothing about it either way — agreed to leave rigorous verification of that edge case out of this PR.

On the `multibox` test: kept, with the benchmark JSON regenerated from the fixed build — the recorded one had come from the un-synced path (the pre-fix build reproduces it to 0.0; the fixed run differs from it by up to 5.6e-3). Your seam input is a different case (cavity ±0.4, theta 0.05, `max_grid_size 8`, 227 steps): at mgs 8 it exercises far more seam borrowing than the current input (36 one-way faces; pre-fix vs fixed differ at O(1)), at mgs 16 it never crosses a seam, and `analysis_fields_2d.py` hard-codes Lx and theta, so it is better added as its own test than swapped in. Happy to do that as a follow-up if you would like it in CI.

@RemiLehe — the split you asked for on Aug 11 is done: #7152 (merged) carries the atomic area accounting and #7153 the convergence tests; this PR is the cross-box core only.

**Comment by aarontran** [comment]
Last two commits + all comment replies LGTM, thanks!

**Comment by ax3l** [comment]
hi @clarkse-he , #7260 landed and this PR needs now a small update pls.
